modules/highgui/cmake/detect_wayland.cmake

```diff
@@ -0,0 +1,68 @@
+# --- Wayland ---
+macro(ocv_wayland_generate protocol_file output_file)
+    add_custom_command(OUTPUT ${output_file}.h
+            COMMAND ${WAYLAND_SCANNER_EXECUTABLE} client-header < ${protocol_file} > ${output_file}.h
+            DEPENDS ${protocol_file})
+    add_custom_command(OUTPUT ${output_file}.c
+            COMMAND ${WAYLAND_SCANNER_EXECUTABLE} private-code < ${protocol_file} > ${output_file}.c
+            DEPENDS ${protocol_file})
+    list(APPEND WAYLAND_PROTOCOL_SOURCES ${output_file}.h ${output_file}.c)
+endmacro()
+
+ocv_clear_vars(HAVE_WAYLAND_CLIENT HAVE_WAYLAND_CURSOR HAVE_XKBCOMMON HAVE_WAYLAND_PROTOCOLS)
+if (WITH_WAYLAND)
+    ocv_check_modules(WAYLAND_CLIENT wayland-client)
+    if (WAYLAND_CLIENT_FOUND)
+        set(HAVE_WAYLAND_CLIENT ON)
+    endif ()
+    ocv_check_modules(WAYLAND_CURSOR wayland-cursor)
+    if (WAYLAND_CURSOR_FOUND)
+        set(HAVE_WAYLAND_CURSOR ON)
+    endif ()
+    ocv_check_modules(XKBCOMMON xkbcommon)
+    if (XKBCOMMON_FOUND)
+        set(HAVE_XKBCOMMON ON)
+    endif ()
+    ocv_check_modules(WAYLAND_PROTOCOLS wayland-protocols>=1.13)
+    if (WAYLAND_PROTOCOLS_FOUND)
+        pkg_get_variable(WAYLAND_PROTOCOLS_BASE wayland-protocols pkgdatadir)
+        find_host_program(WAYLAND_SCANNER_EXECUTABLE NAMES wayland-scanner REQUIRED)
+        ocv_wayland_generate(
+                ${WAYLAND_PROTOCOLS_BASE}/stable/xdg-shell/xdg-shell.xml
+                xdg-shell-client-protocol)
+        set(HAVE_WAYLAND_PROTOCOLS ON)
+    endif ()
+
+    if (HAVE_WAYLAND_CLIENT AND HAVE_WAYLAND_CURSOR AND HAVE_XKBCOMMON AND HAVE_WAYLAND_PROTOCOLS)
+        set(HAVE_WAYLAND TRUE)
+    endif ()
+
+elseif (HAVE_WAYLAND)
+    ocv_add_external_target(wayland
+            "${WAYLAND_CLIENT_INCLUDE_DIRS};${WAYLAND_CURSOR_INCLUDE_DIRS};${XKBCOMMON_INCLUDE_DIRS}"
```

Comment:
This macro creates target `ocv.3rdparty.wayland`.
In general this part looks fine. However target is not really used in `modules/highgui/CMakeLists.txt` (`WAYLAND_CLIENT_LINK_LIBRARIES` is used instead)

---

This macro should be used with autodetection too:

```.diff
-elseif(HAVE_WAYLAND)
+if(HAVE_WAYLAND)
```

---

This could be changed/improved later, e.g. with adding of loadable plugin support.

Reply:
I'm not clear on the usage pattern you are proposing.  I only see ocv.3rdparty.xxx target usage for plugins, so I just removed the usage of ocv_add_external_target.